File: Ruby/lib/problem.rb
class Problem
  
  def initialize(submitted_answer=nil)
    @submitted_answer = nil
    @duration = nil
    yield(self) if block_given?
  end
  
  attr_reader :computed_answer, :submitted_answer
  attr_writer :submitted_answer
  attr_reader :duration
  
  def check(answer, &block)
    computed_answer = block.call
    output_check(computed_answer, answer)
    return computed_answer
  end
  
  def run(&block)
    start_at = Time.now
    @computed_answer = block.call
    stop_at = Time.now
    @duration = stop_at - start_at
    if( submitted_answer )
      output_check(self.computed_answer, self.computed_answer)
    end
    puts "-- Computed in #{self.duration * 1000.0} ms"
    puts "ANSWER: #{submitted_answer.inspect}"
    return @computed_answer
  end
  
  private
  
  def output_check(computed_answer, right_answer)
    message = "-- Answer Correct: #{(computed_answer==right_answer) ? 'YES' : 'NO'}, Expected: #{right_answer.inspect}, Got: #{computed_answer.inspect}"
    puts message
    return message
  end
  
end

File: C/lib/problem.c
#include <stdio.h>
#include <time.h>

#include "problem.h"

/* Declare Private Functions */
void ProblemOutputIntegerCheck(ProblemInteger computedAnswer, ProblemInteger rightAnswer);

/* Do an Integer Check */
bool ProblemIntegerCheck(ProblemInteger answer, ProblemIntegerProcedure *fp) {
	ProblemInteger computedAnswer = (*fp)();
	ProblemOutputIntegerCheck(computedAnswer, answer);
	return computedAnswer == answer;
}

/* Run an Integer Problem */
ProblemInteger ProblemIntegerRun(ProblemInteger submittedAnswer, bool hasSubmittedAnswer, ProblemIntegerProcedure *fp) {
	clock_t startAt = clock();
	ProblemInteger computedAnswer = (*fp)();
	clock_t stopAt = clock();
	double deltaT = (double)(stopAt - startAt) / CLOCKS_PER_SEC;
	if( hasSubmittedAnswer )
		ProblemOutputIntegerCheck(computedAnswer, submittedAnswer);
	printf("-- Computed in %f ms\n", deltaT*1000.0);
	printf("ANSWER: %" PRIPI "\n", computedAnswer);
	return computedAnswer;
}

/* Private function for outputting the check results. */
void ProblemOutputIntegerCheck(ProblemInteger computedAnswer, ProblemInteger rightAnswer) {
	printf("-- Answer Correct: %s, Expected: %" PRIPI ", Got: %" PRIPI "\n", computedAnswer==rightAnswer ? "YES" : "NO", computedAnswer, rightAnswer);
}
